core/src/storage/device.py

```diff
@@ -344,3 +349,29 @@ def get_hide_passphrase_from_host() -> bool:
     Whether we should hide the passphrase from the host.
     """
     return common.get_bool(_NAMESPACE, _HIDE_PASSPHRASE_FROM_HOST)
+
+
+if utils.USE_THP:
+
+    def get_device_secret() -> bytes:
+        """
+        Device secret is used to derive keys that are independent of the seed.
+        """
+        device_secret = common.get(_NAMESPACE, _DEVICE_SECRET)
+        if not device_secret:
+            from trezor.crypto import random
+
+            device_secret = random.bytes(16, True)
+            common.set(_NAMESPACE, _DEVICE_SECRET, device_secret)
+        return device_secret
+
+    def get_cred_auth_key_counter() -> bytes:
+        counter = common.get(_NAMESPACE, _CRED_AUTH_KEY_COUNTER)
+        if counter is None:
+            counter = (0).to_bytes(4, "big")
+            common.set(_NAMESPACE, _CRED_AUTH_KEY_COUNTER, counter)
+        return counter
+
+    def increment_cred_auth_key_counter() -> None:
+        counter = int.from_bytes(get_cred_auth_key_counter(), "big")
+        common.set(_NAMESPACE, _CRED_AUTH_KEY_COUNTER, (counter + 1).to_bytes(4, "big"))
```

Comment:
Is int overflowing here a feature or a bug?

Reply:
Well, it is a bit of both.

The `increment_cred_auth_key_counter` function is called only when user wants to invalidate all credentials issued by Trezor, but not "wipe the storage". The counter has 4 bytes. I cannot imagine any "normal" scenario when this overflow is an expectable outcome. 

On the other hand, there should be a defined behavior for this unlikely scenario. I would like to know the opinions of @matejcik and @andrewkozlik . An argument against the current "silent overflow" is that it could make previously invalidated credentials valid again (until a `device_secret` is re-generated). However, it is unclear to me how to handle a potential overflow error - what should be the message to the user? If overflow happens, we could forbid user to invalidate credentials without also invalidating (re-generating) the `device_secret` - but it seems like a rather complicated UX.

Reply:
Python's `to_bytes()` would raise an overflow exception, but micropython doesn't. Since this operation requires user interaction, there is no way that it is going to overflow under any kind of normal usage conditions. I'd raise the `OverflowError` "manually", or better yet, use `utils.ensure()` to protect against malicious scenarios.

Reply:
Btw I was thinking about it before ...

wouldn't it make sense to introduce a function `common.inc(namespace, variable)` that does the increment and checks for overflows?

Reply:
utils.ensure added in [5dce7cc](https://github.com/trezor/trezor-firmware/pull/3861/commits/5dce7cc836524bdc93f590bb8915c7098e3d1d76)

Reply:
test for overflow added in [8a4b39d](https://github.com/trezor/trezor-firmware/pull/3861/commits/8a4b39da4eafc5bf69cf3caf7ab4aee5f6d04d34)

Reply:
@matejcik what is your opinion on @prusnak 's proposal to introduce `common.inc` function, do you think it should be a part of this PR?

Reply:
we have just this one counter though, don't we? i'd add `common.inc` only if we ever implement more

Reply:
Well, there is for instance `U2F_COUNTER`. However, it is used as an `int` and there is no need to access the "current counter value". It has functions `set_u2f_counter` and `next_u2f_counter` implemented in the `device.py`.

IMO I would keep it as it is now.

Reply:
FTR, there is a big difference between the U2F counter and this one in that the U2F counter is not encrypted, which means it can use an optimized implementation in bitwise-writable flash storage such as that on the STM32F series. (The STM32U5 flash is not bitwise-writable.) This optimization is worthwhile only for counters that are incremented often, e.g. with every read.